File: android/src/main/java/com/evotor/modules/EventModule.java
package com.evotor.modules;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import com.evotor.services.EventListenerService;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by a.lunkov on 21.12.2017.
 */

public class EventModule extends ReactContextBaseJavaModule {

    private static boolean mounted = false;
    private static List<String> listenersEnabled = new ArrayList<>();
    private static ActivityInspector activityInspector = null;

    public EventModule(final ReactApplicationContext c) {
        super(c);
        activityInspector = new ActivityInspector() {
            @Override
            public boolean isActivityRunning() {
                return getCurrentActivity() != null;
            }
        };
    }

    @Override
    public final String getName() {
        return "EventModule";
    }

    @Override
    public final Map<String, Object> getConstants() {
        return new HashMap<>();
    }

    @ReactMethod
    public void setListenerEnabled(final String eventName, final boolean enabled) {
        if (enabled) {
            if(!listenersEnabled.contains(eventName)) {
                listenersEnabled.add(eventName);
            }
        } else if (listenersEnabled.contains(eventName)) {
            listenersEnabled.remove(eventName);
        }
    }

    public static Intent createServiceIntent(final Context context, final String eventName, final boolean isCustom) {
        if (!mounted) {
            context.startService(new Intent(context, EventListenerService.class));
            mounted = true;
        }
        if (!isCustom && !listenersEnabled.contains(eventName)) {
            return null;
        }
        Intent result = new Intent(context, EventListenerService.class);
        result.putExtra("type", eventName);
        result.putExtra("appIsRunning", activityInspector != null && activityInspector.isActivityRunning());
        return result;
    }

    public static boolean startService(final Context context, final String eventName, final Bundle data) {
        Intent service = createServiceIntent(context, eventName, false);
        if(service == null) {
            return false;
        }
        if(data != null) {
            service.putExtra("extras", data);
        }
        context.startService(service);
        return true;
    }

    private interface ActivityInspector {
        boolean isActivityRunning();
    }

}

File: android/src/main/java/com/evotor/modules/EventModule.kt
package com.evotor.modules

import android.content.Context
import android.content.Intent
import android.os.Bundle

import com.evotor.converter.from.js.NavigationReader
import com.evotor.converter.to.js.EventWriter
import com.evotor.services.EventEmissionService
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule

import java.util.HashMap

/**
 * Created by a.lunkov on 21.12.2017.
 */

class EventModule(private val context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {

    init {
        localEmitter = object : LocalEmitter {
            override fun emit(data: WritableMap) {
                context
                        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                        .emit("LocalEventEmission", data)
            }
        }
    }

    override fun getName(): String {
        return "EventModule"
    }

    override fun getConstants(): Map<String, Any> {
        return HashMap()
    }

    internal interface LocalEmitter {
        fun emit(data: WritableMap)
    }

    companion object {

        private var localEmitter: LocalEmitter? = null

        fun emit(c: Context, eventName: String, data: Any?) {
            if (!emitLocal(eventName, data)) {
               emitGlobal(c, eventName, data)
            }
        }

        fun emitLocal(eventName: String, data: Any?): Boolean {
            localEmitter?.emit(EventWriter.writeLocalEvent(eventName, data))
            return localEmitter != null
        }

        fun emitGlobal(c: Context, eventName: String, data: Any?) {
            val extras = Bundle()
            NavigationReader.readDefaultIntentExtra(c, "extras", data, extras)
            c.startService(
                    Intent(c, EventEmissionService::class.java)
                            .putExtra("type", eventName)
                            .putExtras(extras)
            )
        }

    }

}
